modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/components/Chart.js

```diff
@@ -25,9 +25,18 @@ import {
 	YAxis,
 } from 'recharts';
 
+import {
+	useAddDataSetItems,
+	useChangeTimeSpanKey,
+	useChartState,
+	useDateTitle,
+	useIsPreviousPeriodButtonDisabled,
+	useNextTimeSpan,
+	usePreviousTimeSpan,
+	useSetLoading,
```

Comment:
Interesting choice to structure this using so many hooks. My general rule of thumb is that hooks are all about _reuse_, so if a hook is only used in exactly one place, I wouldn't make it a custom hook. It seems that most of these are once-only hooks.

I think it's fine if you want to keep it this way, so just planting this as food for thought. The real trade-off here is because the hook custom hook is a layer of indirection, of abstraction, that conceals what's really happening. If "what's really happening" is verbose and painful to duplicate, the hook provides us with a way to encapsulate the functionality in a small unit (ie. reuse). If "what's really happening" is complex and difficult to understand, the hook can make the system more intellectually manageable, although for this to be true the concepts and naming have to be clear, otherwise it might be a lateral step at best. Conversely, if it is _not_ complex and difficult to understand, then the indirection might end up obfuscating the structure and relationships in the data model, and that could be information that would be useful to have in people's mental models.

In other words, as always in engineering, a set of trade-offs. But as a general starting point, the "hooks are for reuse" rule of thumb has served me pretty well.

Reply:
I totally agree with you and that's the principal idea of this pull request. 

I created those custom hooks because the idea is to make it reusable in almost every detail view and also decouple from the `Chart` component. We are actually working on Referral (see https://github.com/liferay-frontend/liferay-portal/pull/573, you can give feedback also 😛) and Social detail view (still WIP) from traffic channels. In those views, we have the same selector (7 Days, 30 Days) as in the first level and we want to share the state between those screens (main-detail).

Looking at that list once again, I think that the only one that maybe will be used in one component is the `useAddDataSetItems` (do you remember it right? This method converts the data to an understandable one for the chart, you helped me with this 😄 ). The others, as I said, will be used in detail view.

Reply:
So here's the other axis to think about, beyond re-use: if the same set of N hooks are used in multiple places, maybe they should just be one hook (or at least `< N` hooks); in this case, a few of them are just wrappers around `useContext`, that return `dispatch`, so "the one hook" might just be `useContext`. I'm not saying you should change this, but just wanted to plant it as a seed of an idea: that if you need to use a dozen custom hooks many times over to get something done, you may not have actually made the system easier to use overall (especially the cases where the hooks are internally simple: because the indirection just means that you have to jump to another place in order to see what they actually do, and you also have to remember what they all are etc).

Reply:
I'll take into account for next implementations. Thanks :)

Reply:
See the use of these custom hooks in ReferralDetail: https://github.com/liferay-frontend/liferay-portal/pull/582 (2 commits). Sooner or later it seems that I will have a unique custom hook for these `N` actions 🤔